sdk/ai/Azure.AI.Inference/src/Telemetry/OpenTelemetryScope.cs

```diff
@@ -0,0 +1,247 @@
+            var errorType = e?.GetType()?.FullName;
+            if ( e is Azure.RequestFailedException requestFailed)
+            {
+                errorType = requestFailed.Status.ToString();
+            }
+            m_activity.SetTag(ErrorTypeKey, errorType);
+            m_activity.SetStatus(ActivityStatusCode.Error, e?.Message ?? errorType);
+        }
+
+        /// <summary>
+        /// Record the events and metrics associated with the response.
+        /// </summary>
+        /// <param name="recordedResponse"></param>
+        private void RecordResponseInternal(AbstractRecordedResponse recordedResponse)
+        {
+            if (m_activity == null)
+                return;
+            TagList tags = m_commonTags;
+            // Find index of model tag.
+            object objOldModel = m_activity.GetTagItem(GenAiResponseModelKey);
+            if (objOldModel != null)
+            {
+                tags.Remove(new KeyValuePair<string, object> (GenAiResponseModelKey, objOldModel));
+            }
+            tags.Add(GenAiResponseModelKey, recordedResponse.Model);
+            // Record duration metric
+            s_duration.Record(m_duration.Elapsed.TotalSeconds, tags);
+            // Record input tokens
+            if (recordedResponse.PromptTokens != AbstractRecordedResponse.NOT_SET)
+            {
+                TagList input_tags = tags;
+                input_tags.Add(GenAiUsageInputTokensKey, "input");
+                s_tokens.Record(recordedResponse.PromptTokens, input_tags);
+                m_activity.SetTag(GenAiUsageInputTokensKey, recordedResponse.PromptTokens);
+            }
+            // Record output tokens
+            if (recordedResponse.CompletionTokens != AbstractRecordedResponse.NOT_SET)
+            {
+                TagList output_tags = tags;
+                output_tags.Add(GenAiUsageOutputTokensKey, "output");
+                s_tokens.Record(recordedResponse.CompletionTokens, output_tags);
+                m_activity.SetTag(GenAiUsageOutputTokensKey, recordedResponse.CompletionTokens);
+            }
+
+            // Record the event for each response
+            string[] choices = recordedResponse.GetSerializedCompletions();
+            foreach (string choice in choices)
+            {
+                TagList completionTags = new()
+                {
+                    { GenAiSystemKey, GenAiSystemValue},
+                    { GenAiEventContent, JsonSerializer.Serialize(choice) }
+                };
+                m_activity.AddEvent(new ActivityEvent(
+                        GenAiChoice,
+                        DateTimeOffset.Now,
+                        new ActivityTagsCollection(completionTags)
+                ));
+            }
+            // Set activity tags
+            if (!string.IsNullOrEmpty(recordedResponse.FinishReason))
+            {
+                m_activity.SetTag(GenAiResponseFinishReasonKey, recordedResponse.FinishReason);
```

Comment:
`recordedResponse.FinishReason` is it a string? We need to provide a string here, otherwise it may be dropped or the enum constant name will be used instead of `stop`/`tool_calls`/etc - this one should be the same across different languages.

Reply:
In Azure.AI.Inference the CompletionsFinishReason is an IEquitable, which essentially stores the string literals for finish reasons:
https://github.com/Azure/azure-sdk-for-net/blob/0f72179563faacabc7a93838883b34a092854213/sdk/ai/Azure.AI.Inference/src/Generated/CompletionsFinishReason.cs#L22

These literals are the same as in semantic convntions, though we do not have the "function_call" reason.